File: lld/problems/category_b_machine_coding/incpp/in_memory_job_scheduler/service/jobSchedulerService.cpp
#include"../dao/jobDao.cpp" 
#include<chrono> 
#include<iostream>
#include<queue> 
#include<mutex>
#include<thread>
using namespace std; 



class JobSchdulerServiceI {
    public: 
        virtual void scheduleImmediate(string jobData, chrono::seconds ts ) = 0;  // later on error handling
        virtual void schedule(string jobData, int period ) = 0; 
        virtual void run() = 0;
};


class JobSchdulerService: public JobSchdulerServiceI {
    JobDao* jobDao; 
    int workerThreads; 
    mutex mu; 
    condition_variable cv;
    queue<ExecutionJob> workerQueue; 
    const int period = 20;
    void organiser() {
        while(true) {
            cout<<"........................"<<endl;
            cout<<"organiser started"<<endl;
            chrono::steady_clock::time_point st = chrono::steady_clock::now();
            chrono::steady_clock::time_point et = chrono::steady_clock::now() + chrono::seconds(period);
            vector<ExecutionJob> jobs =  this -> jobDao -> getExJobRange(st, et);
            for(ExecutionJob job: jobs) {
                job.updateStatue(JobStatus::SCHEDULED);   
                
                {
                    lock_guard<mutex> lock(mu);
                    this -> workerQueue.push(job);

                }
                this -> jobDao -> updateExJob(job);
                cv.notify_all();
            }
            cout<<"organiser sleeping"<<endl;
            cout<<"........................"<<endl;
            this_thread::sleep_for(chrono::seconds(10));
        }
        

    }
    void scheduler() {
        while(true){
            cout<<"........................"<<endl;
            cout<<"scheduler started"<<endl;
            chrono::steady_clock::time_point st = chrono::steady_clock::now();
            chrono::steady_clock::time_point et = chrono::steady_clock::now() + chrono::seconds(period);
            vector<ScheduleJob> jobs =  this -> jobDao -> getScJobRange(st, et);
            for(ScheduleJob job: jobs) {
                JobI* jobPtr = this->jobDao->getJob(job.getJobId(), JobType::RECURRING);
                ExecutionJob ej = ExecutionJob(jobPtr, job.nextTs);
                job.nextTs = job.nextTs + chrono::seconds(this -> period);
                this -> jobDao -> updateScJob(job);
                this -> jobDao -> updateExJob(ej);
            }
            cout<<"scheduler sleeping"<<endl;
            cout<<"........................"<<endl;
            this_thread::sleep_for(chrono::seconds(10));
        }
        
    }
    void worker() {
        while(true){
            unique_lock<mutex> lock(mu);
            cout<<"........................"<<endl;
            cout<<"worker waiting"<<endl;
            cv.wait(lock, [this](){ return this -> workerQueue.size() > 0 ;});
            cout<<"worker starting"<<endl;
            if (!this->workerQueue.size()) continue;
            ExecutionJob j = this -> workerQueue.front();
            this -> workerQueue.pop();
            cout<<"working execution job with executionId "<<j.getJobId()<<" "<<j.getData()<<endl; 
            j.status  = JobStatus::SUCCESS;
            this -> jobDao->updateExJob(j);
            cout<<"........................"<<endl;
        }
        
    }
    public: 
        JobSchdulerService(){}
        JobSchdulerService(JobDao* jd, int wt): jobDao(jd), workerThreads(wt) {}
        void scheduleImmediate(string jobData, chrono::seconds ts ) override {
            ImmediateJob cj = this -> jobDao->createImmediateJob(ImmediateJob(jobData, ts), JobType::NON_RECURRING);
            chrono::steady_clock::time_point ct = chrono::steady_clock::now() + chrono::seconds(ts);
            // auto seconds_since_epoch = chrono::duration_cast<chrono::seconds>(ct.time_since_epoch()).count();
            // cout<<seconds_since_epoch<<endl;
            JobI* jobPtr = this->jobDao->getJob(cj.getJobId(), JobType::NON_RECURRING);
            ExecutionJob ej = ExecutionJob(jobPtr, ct);
            this -> jobDao -> createExecutionJob(ej);
            return ;
        }   
        void schedule(string jobData,  int period  ) override{
            ScheduleJob sj = ScheduleJob(jobData, period);
            this -> jobDao -> createSchedule(sj);
            return ; 
        }

        void run() override{
            thread t1(&JobSchdulerService::organiser, this);
            thread t2(&JobSchdulerService::scheduler, this);
            thread t3(&JobSchdulerService::worker, this);
            t1.join();
            t2.join();
            t3.join();
        }
};

File: lld/problems/category_b_machine_coding/ingolang/in_memory_job_scheduler_go/service/jobSchedulerService.go
package service

import (
	"context"
	"fmt"
	"sync"
	"time"
	"in_memory_job_scheduler_go/dao"
	"in_memory_job_scheduler_go/models"
)

type JobSchedulerServiceI interface {
	ScheduleImmediate(jobData string, ts time.Duration)
	Schedule(jobData string, period int)
	Run(ctx context.Context)
}

type JobSchedulerService struct {
	jobDao        *dao.JobDao
	workerThreads int
	workerQueue   chan *models.ExecutionJob
	period        time.Duration
}

func NewJobSchedulerService(jd *dao.JobDao, workerThreads int) *JobSchedulerService {
	return &JobSchedulerService{
		jobDao:        jd,
		workerThreads: workerThreads,
		workerQueue:   make(chan *models.ExecutionJob, 100),
		period:        20 * time.Second,
	}
}

func (s *JobSchedulerService) organiser(ctx context.Context) {
	ticker := time.NewTicker(10 * time.Second)
	defer ticker.Stop()
	
	for {
		select {
		case <-ctx.Done():
			return
		case <-ticker.C:
			fmt.Println("organiser started")
			st := time.Now()
			et := st.Add(s.period)
			jobs := s.jobDao.GetExJobRange(st, et)
			
			for _, job := range jobs {
				job.UpdateStatus(models.SCHEDULED)
				s.jobDao.UpdateExJob(job)
				select {
				case s.workerQueue <- job:
				case <-ctx.Done():
					return
				}
			}
			fmt.Println("organiser sleeping")
		}
	}
}

func (s *JobSchedulerService) scheduler(ctx context.Context) {
	ticker := time.NewTicker(10 * time.Second)
	defer ticker.Stop()
	
	for {
		select {
		case <-ctx.Done():
			return
		case <-ticker.C:
			fmt.Println("scheduler started")
			st := time.Now()
			et := st.Add(s.period)
			jobs := s.jobDao.GetScJobRange(st, et)
			
			for _, job := range jobs {
				if jobPtr := s.jobDao.GetJob(job.Id, models.RECURRING); jobPtr != nil {
					ej := models.NewExecutionJob(jobPtr, job.NextTs)
					job.NextTs = job.NextTs.Add(s.period)
					s.jobDao.UpdateScJob(job)
					s.jobDao.UpdateExJob(ej)
				}
			}
			fmt.Println("scheduler sleeping")
		}
	}
}

func (s *JobSchedulerService) worker(ctx context.Context) {
	for {
		select {
		case <-ctx.Done():
			return
		case job := <-s.workerQueue:
			fmt.Printf("working execution job id=%d data=%s\n", job.Id, job.GetData())
			job.UpdateStatus(models.SUCCESS)
			s.jobDao.UpdateExJob(job)
		}
	}
}

func (s *JobSchedulerService) ScheduleImmediate(jobData string, ts time.Duration) {
	cj := s.jobDao.CreateImmediateJob(models.NewImmediateJob(jobData, ts))
	ej := models.NewExecutionJob(cj, time.Now().Add(ts))
	s.jobDao.CreateExecutionJob(ej)
}

func (s *JobSchedulerService) Schedule(jobData string, period int) {
	s.jobDao.CreateSchedule(models.NewScheduleJob(jobData, period))
}

func (s *JobSchedulerService) Run(ctx context.Context) {
	var wg sync.WaitGroup
	wg.Add(3)
	
	go func() {
		defer wg.Done()
		s.organiser(ctx)
	}()
	
	go func() {
		defer wg.Done()
		s.scheduler(ctx)
	}()
	
	go func() {
		defer wg.Done()
		s.worker(ctx)
	}()
	
	wg.Wait()
}

